### PR TITLE
[LICM] Prevent LICM of ptrtoint and inttoptr when using non-integral pointers

### PR DESCRIPTION
**Pull request by Il-Capitano**
Non-integral pointers don't have a stable bit representation, i.e. subsequent executions of `ptrtoint` or `inttoptr` instructions can create different results, therefore it is invalid to move these instructions outside of a loop when they use non-integral pointer types.

### CODE REVIEW

**Review by DanielKristofKiss — APPROVED** [review]
LGTM, could you rerun clang-format?

**Review by nikic — COMMENTED** [review]
I'm pretty sure that this was not an indented consequence of non-integral pointers. Please do not merge this until an expert on non-integral pointers has approved this.

File: llvm/test/Transforms/LICM/non-integral-pointers.ll
Comment (nikic) [inline llvm/test/Transforms/LICM/non-integral-pointers.ll:1]: Use update_test_checks.py.

Comment (Il-Capitano) [reply llvm/test/Transforms/LICM/non-integral-pointers.ll:2]: Done.

**Comment by Il-Capitano** [comment]
@preames @arpilipe @annamthomas Pinging for review, please.

**Comment by Il-Capitano** [comment]
@preames @arpilipe @annamthomas Ping

**Comment by arpilipe** [comment]
There was a bit of discussion about LICM in the review that allowed such casts for non-integral pointers.
https://reviews.llvm.org/D104547

At that time, we considered LICM valid because the cast is allowed to produce the same result every time. 

What kind of problem are you trying to solve with this restriction?

**Comment by Il-Capitano** [comment]
In my use case, the result of `ptrtoint` is used for indexing a GC card table. The issue was that during the loop's execution, the original pointer could get relocated, meaning that a different index would be needed for the card table marking potentially every iteration. However, LICM hoisted the `ptrtoint` outside of the loop, resulting in the same index being used every iteration, which caused an issue.

Looking at the discussion you linked, and the [wording in the Language Reference](https://llvm.org/docs/LangRef.html#nointptrtype), I guess my question is what is meant by *implementation defined fencing*? I assumed that having a call not marked with `"gc-leaf-function"` (or at least `memory(inaccessiblemem: write)` to allow it changing the environment state) in the execution path would have sufficed, but LICM (and maybe other optimizations?) doesn't respect that currently. Would this fencing rather be, that the `ptrtoint` should be replaced by an external function call in order to observe the pointer's bit representation at that particular program state?

**Comment by arpilipe** [comment]
Here is how we implement fencing: Operations that inspect the bitwise representation of GC pointers are wrapped into function calls. These calls remain opaque to the optimizer until they are inlined. We inline them only after we have inserted explicit relocations using RewriteStatepointsForGC. Once the relocations are explicit, statepoints act as barriers to optimizations.

For example, in your case, if a loop contains a statepoint, the corresponding relocation will create a new def for the value, preventing LICM.

Restricting the hoisting of inttoptr and ptrtoint for non-integral pointers will prevent hoisting in cases where the appropriate fencing has been done. E.g., in our scenario, hoisting is possible after RewriteStatepointsForGC if the loop doesn't contain statepoints, thus can't change the bitwise representation of the pointers.

I'm not sure if this has a significant performance impact, so I don't have a strong preference. However, before we restrict the optimization, you may consider a similar fencing scheme to prevent undesirable transformations.

**Comment by Il-Capitano** [comment]
Thanks for the explanation, using an opaque function before relocations are finalized seems like the best approach, I'll try that.

**Comment by arichardson** [comment]
I recently started some work trying to clarify the non-integral pointer semantics. Right now we assume non-integral pointers are both numerically unstable (e.g. copying GC) as well as not just an address. I'd like to split these two properties (Draft PR https://github.com/llvm/llvm-project/pull/105735) and then we'd be able to restrict this to unstable pointers while still allowing it for fat pointers and/or CHERI capabilities.

**Comment by Il-Capitano** [comment]
I'm closing this PR, as the intended solution seems to be to use a noinline function in place of pointer casts before relocations are finalized. I also realized that if we wanted to make `inttoptr` and `ptrtoint` follow similar semantics with non-integral pointers, other passes need to be changed as well, not just LICM, e.g. https://godbolt.org/z/ejqzbos7f.

**Pull request closed without merging**
